packages/react-native-editor/__device-tests__/pages/editor-page.js

```diff
@@ -542,35 +543,70 @@ class EditorPage {
 		return await blockLocator.text();
 	}
 
+	async getNumberOfParagraphBlocks() {
+		const paragraphBlockLocator = isAndroid()
+			? `//android.view.ViewGroup[contains(@content-desc, "Paragraph Block. Row")]/android.widget.EditText`
+			: `(//XCUIElementTypeButton[contains(@name, "Paragraph Block. Row")])`;
+		const locator = await this.driver.elementsByXPath(
+			paragraphBlockLocator
+		);
+		return locator.length;
+	}
+
+	async assertSlashInserterPresent() {
+		const slashInserterLocator = isAndroid()
+			? '//android.widget.HorizontalScrollView[@content-desc="Slash inserter results"]/android.view.ViewGroup'
+			: '(//XCUIElementTypeOther[@name="Slash inserter results"])[1]';
+
+		return await isElementVisible( this.driver, slashInserterLocator, 5 );
+	}
+
 	// =========================
 	// List Block functions
 	// =========================
 
-	async getTextViewForListBlock( block ) {
-		let textViewElementName = 'XCUIElementTypeTextView';
-		if ( isAndroid() ) {
-			textViewElementName = 'android.widget.EditText';
-		}
+	async getListBlockAtPosition(
+		position = 1,
+		options = { isEmptyBlock: false }
+	) {
+		// Go to the correct list at position
+		const listBlockPositionLocator = isAndroid()
+			? `//android.view.ViewGroup[@content-desc="List Block. Row ${ position }"]`
+			: `(//XCUIElementTypeOther[contains(@name, "List Block. Row ${ position }")])[5]`;
 
-		const accessibilityId = await block.getAttribute(
-			this.accessibilityIdKey
+		let listBlock = await waitForVisible(
+			this.driver,
+			listBlockPositionLocator
 		);
-		const blockLocator = `//*[@${
-			this.accessibilityIdXPathAttrib
-		}=${ JSON.stringify( accessibilityId ) }]//${ textViewElementName }`;
-		return await this.driver.elementByXPath( blockLocator );
-	}
+		await listBlock.click();
+
+		// iOS needs a click to get the text element
+		if ( ! isAndroid() ) {
+			const listBlockLocator = options.isEmptyBlock
+				? `(//XCUIElementTypeStaticText[contains(@name, "List")])`
+				: `//XCUIElementTypeButton[contains(@name, "List")]`;
+
+			listBlock = await waitForVisible( this.driver, listBlockLocator );
+			await listBlock.click();
+		}
 
-	async sendTextToListBlock( block, text ) {
-		const textViewElement = await this.getTextViewForListBlock( block );
+		const listBlockTextLocatorIOS = options.isEmptyBlock
+			? `(//XCUIElementTypeStaticText[contains(@name, "List")])`
+			: `//XCUIElementTypeButton[contains(@name, "List")]//XCUIElementTypeTextView`;
```

Comment:
This locator looks very similar to the one defined in lines 585 - 587, I'm wondering if we could reduce the code here by unifying both parts. The only difference I spotted is that when the block is not empty, here we're adding `//XCUIElementTypeTextView` at the end of the xpath, is this required? If so, then I understand that we need both locators.

https://github.com/WordPress/gutenberg/blob/85c407c286fc18e226f98fbcfa440237f995d093/packages/react-native-editor/__device-tests__/pages/editor-page.js#L585-L587

Reply:
Ah yes, this is needed for list blocks, the locator for when the list is empty vs not empty is different. For both cases on iOS, we'll need to click on it first (lines 584 - 591). After clicking on it, when it's empty the XPath remains the same but when it's not empty, it will need `//XCUIElementTypeTextView` to get the correct element (lines 593 - 595).

Reply:
Ok, thanks for the clarification, in that case, let's continue having the two locators 👍 .